vmm/src/vm.rs

```diff
@@ -1232,17 +1291,35 @@ impl<'a> Vm<'a> {
             Some(file) => {
                 let mut mem_regions = Vec::<(GuestAddress, usize, Option<FileOffset>)>::new();
                 for region in ram_regions.iter() {
-                    let file = OpenOptions::new()
-                        .read(true)
-                        .write(true)
-                        .custom_flags(O_TMPFILE)
-                        .open(file)
-                        .map_err(Error::SharedFileCreate)?;
-
-                    file.set_len(region.1 as u64)
-                        .map_err(Error::SharedFileSetLen)?;
+                    if file.is_file() {
+                        let file = OpenOptions::new()
+                            .read(true)
+                            .write(true)
+                            .open(file)
+                            .map_err(Error::SharedFileCreate)?;
+
+                        file.set_len(region.1 as u64)
+                            .map_err(Error::SharedFileSetLen)?;
+
+                        mem_regions.push((region.0, region.1, Some(FileOffset::new(file, 0))));
+                    } else if file.is_dir() {
+                        let fs = format!("{}{}", file.display(), "/tmpfile_XXXXXX");
+                        println!("fs is {}", fs);
+                        let fp = Path::new(&fs);
+                        let fd = match unistd::mkstemp(fp) {
```

Comment:
Can you look at at alternative that we might already be using? tempfile?

Reply:
Do you mean the following way?

OpenOptions::new()
                    .read(true)
                    .write(true)
                    .custom_flags(O_TMPFILE)
                    .open(pmem_cfg.file)

O_TMPFILE requires support by the underlying filesystem; only a subset of Linux filesystems provide that support, like ex2, ext3, ext4, UDF, Minix, and shmem filesystems. But it is not implemented for hugetlbfs. That's why I choose the other way.

Reply:
I see tempdir::TempDir in integration test, this is used to format a socket path, no open file operation. The same, I also use format! here to construct the path str.

Reply:
Can you access those functions though the libc crate directly?

Reply:
Ditto!